File: JS/mol/mol.js
#!/usr/bin/env node

const COEFFRE = /^(\d+)/g; // beginning coefficient
const TOKRE = /\(.*?\)|([A-Z][a-z]*)(\d*)/g; // groups: Elem, [coeff]
const SUBRE = /\((.*)\)(\d*)/g; // groups: expr, [coeff]

ptable = {
    H: 1,
    C: 12,
    undefined: 0
}

function makeCoeff(c) {
    if (c == '' || c === undefined) {
        return 1;
    } else {
        return parseInt(c);
    }
}

//count = 0
function mass(thing) {
    console.log(`mass(${thing})`)
    //count++
    //if (count == 2) console.assert(0); // deliberate fail condition
    var acc = 0;
    var big = COEFFRE.exec(thing);
    var bigCoeff = 1;
    if (big !== null) {
        bigCoeff = makeCoeff(big[1]);
    }

    while ((m = TOKRE.exec(thing)) !== null) {
        let toks = m.slice(1, 3);
        acc += ptable[toks[0]] * makeCoeff(toks[1]);
    }

    // problem section
    while ((m = SUBRE.exec(thing)) !== null) {
        let toks = m.slice(1, 3);

        // Causing infinite looping
        //acc += mass(toks[0]) * makeCoeff(toks[1]);
    }

    return bigCoeff * acc;
}

function translate(source, froms, tos) {
    var table = {};
    var ret = [];
    if (froms.length != tos.length) {
        throw "'froms' & 'tos' must be equal length"
    }

    // trans = str.maketrans ...
    for (const i in froms) {
        table[froms[i]] = tos[i];
    }

    // str.translate ...
    for (const c of source) {
        if (table[c] == undefined) {
            ret.push(c);
        } else {
            ret.push(table[c]);
        }
    }

    return ret.join('');
}

function sanitize(thing) {
    thing = translate(thing, "{[]}", "(())");
    var arr = thing.split('');
    arr = arr.filter(c => c.match(/[a-z0-9\(\)]/i));
    return arr.join('');
}

// if(__name__ == '__main__') eqiv, supporting web embedding
if (typeof require != 'undefined' && require.main == module) {
    test = "CH3(CH3)2CH3";
    console.log(mass(test));
    //console.log(sanitize("C224[](){} \t<>?££$£"));
    //console.log(mass("CH3"))
}

File: Python/mol/mol.py
#!/usr/bin/env python3
import re
from pathlib import Path
from csv import reader
COEFFRE = r"^(\d+)"  # beginning coefficient
ELEM = r"([A-Z][a-z]*)"
TOKRE = r"\(.*?\)|" + ELEM + r"(\d*)"  # groups: Elem, [coeff]
SUBRE = r"\((.*)\)(\d*)"  # groups: expr, [coeff]
TRANS = str.maketrans("{}[]", "()()")


def make_coeff(c):
    if c == "":
        return 1
    else:
        return int(c)


def load_table(fname="ptable.tsv"):
    tpath = Path(__file__).resolve().parent / fname
    global ptable
    ptable = {'': 0.00}
    with open(tpath, 'r') as f:
        next(f)  # skip header
        for i, (k, v) in enumerate(reader(f, delimiter='\t')):
            if k in ptable:
                raise Exception(f"Table error: line {i}, Repeated element {k}")
            else:
                ptable[k] = float(v)


load_table()


def valid(c):
    return c.isalnum() or c == "(" or c == ")"


def sanitize(comp):
    comp = comp.translate(TRANS)
    comp = [c for c in comp if valid(c)]
    return ''.join(comp)


class ElementError(Exception):
    def __init__(self, elem):
        super().__init__()
        self.elem = elem

    def __repr__(self):
        return f"Element \"{self.elem}\" doesn't exist"


def mass(comp):
    big = re.match(COEFFRE, comp)
    if big:
        bigCoeff = int(big[0])
    else:
        bigCoeff = 1

    acc = 0
    for e, c in re.findall(TOKRE, comp):
        if e not in ptable:
            raise ElementError(e)
        coeff = make_coeff(c)
        acc += ptable[e] * coeff

    for e, c in re.findall(SUBRE, comp):
        coeff = make_coeff(c)
        acc += mass(e) * coeff

    return bigCoeff * acc


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(
        description="Calculate molecular masses from structural formulae")
    parser.add_argument("comps", metavar="Compound", nargs='+',
                        help="Structural formula to be evaluated")
    args = parser.parse_args()

    for comp in args.comps:
        clean = sanitize(comp)
        try:
            print(f"{clean}:\t{mass(clean):.2f} g/mol")
        except ElementError as e:
            print("%r" % e)
            exit(1)
